[Design] 진행률 셀 추가

taek0622 opened the pull request:
## Motivation 🥳 (코드를 추가/변경하게 된 이유)
- 시공내역에서 상단에 카테고리별 진행률을 보여주기 위한 셀을 추가하기 위함

## Key Changes 🔥 (주요 구현/변경 사항)
- 카테고리별 진행률을 보여주기 위한 `ProgressCell` 추가
- HistoryViewController에 시공내역, 문의내역을 분기하기 위한 `isWorkingView` 프로퍼티 추가
- 시공내역, 문의내역에 따라 컬렉션 뷰의 진행률, AS 셀을 분기하기 위한 구문 추가
- 특정 카테고리 진행률 클릭시 선택된 카테고리만 선택되었다는 느낌으로 보여주는 기능 추가
- 진행률을 물결 형태로 보여주기 위한 WaveView 추가

## ToDo 📆 (남은 작업)
- [ ] 없음

## ScreenShot 📷 (참고 사진)
<img src="https://user-images.githubusercontent.com/81027256/218607803-8d81cdbb-5967-4ead-8bbc-72d87270147a.gif">


## To Reviewers 🙏 (리뷰어에게 전달하고 싶은 말)
- 상당히 간만에 클라이언트를 만져보네요. 드디어 클라이언트를 할 수 있어서 너무 기쁩니다..!
- 사진상으로 눈에 띄지 않을 수 있지만, 현재 Opacity가 적용되어 있는 상태입니다. 선택된 아이템만 opacity가 Hidden되고 있습니다. 해당 부분은 현재 색상이 전체적으로 밝아서 생기는 문제로, 후에 배경색과 차트 색이 바뀌면 눈에 띄게 바뀌게 될 것 같습니다.
- 테두리 선이 겹치는 부분이 굵기가 일정해보이지 않는 문제가 있는데, 해당 부분에 대해 어떻게 고쳐야할지 아직 감이 잘 오지 않습니다. 해당 부분에 대해서는 새 PR에서 고쳐서 올려보겠습니다.
- ~~진행률 부분을 물결선으로 적용해서 올리고 싶었으나, 아직 BezierPath 사용에 미숙한 관계로 우선 물결선을 제외하고 올립니다. 해당 부분은 계속 시도 중이니 완성되는대로 해당 부분에 대한 새로운 PR 올리겠습니다.~~ => WaveView 2023.02.14 09:35 부로 추가 되었습니다.
- 현재 WaveView가 들어갈만한 곳이 마땅치 않아서, cell의 일부이니 cell 디렉토리에 들어가있는데, 해당 부분에 대해 의견이 있다면 남겨주시면 감사하겠습니다.
- 잘 부탁드립니다. 감사합니다.

## Reference 🔗
- [개발자 키우기 - [iOS] CollectionView 선택한 아이템 UI 변경하기](https://superwony.tistory.com/158)
- [Fomagran - [iOS/UI] UIBezierPath의 거의 모든 것 (feat. 모양,곡선,직선 그리기)](https://fomaios.tistory.com/entry/iOSUI-UIBezierPath%EC%9D%98-%EA%B1%B0%EC%9D%98-%EB%AA%A8%EB%93%A0-%EA%B2%83-feat-%EB%AA%A8%EC%96%91%EA%B3%A1%EC%84%A0%EC%A7%81%EC%84%A0-%EA%B7%B8%EB%A6%AC%EA%B8%B0)

## Close Issues 🔒 (닫을 Issue)
- Close #39.

taek0622 committed:
Design:iphone:: #39 Add Progress Cell Layout

- 진행률 셀 레이아웃 추가

taek0622 committed:
Feat:sparkles:: Add isWorkingView property

- 시공상황과 문의내역에 따라 historyViewController를 분기하기 위한 isWorkingView property 추가

taek0622 committed:
Feat:sparkles:: Add progressCell condition

- 시공상황, 문의내역에 따라 진행률 셀과 AS 기간 셀로 분기하는 기능 추가

taek0622 committed:
Chore:wrench:: Modify InteritemSpacing

- 컨텐츠 부분에서는 InterItem이 적용될 일이 없으므로 0으로 통일

taek0622 committed:
Design:iphone:: #39 Modify text attribute

- 텍스트 색상 및 폰트 사이즈, 굵기 변경

taek0622 committed:
Design:iphone:: #39 Add WaveView Layout

- 진행률을 표기하기 위한 WaveView Layout 추가

taek0622 committed:
Desing:iphone:: #40 Apply waveview

- 진행률 셀에 WaveView 적용

taek0622 committed:
Design:iphone:: #39 Apply waveview

- 진행률 셀에 WaveView 적용

taek0622 committed:
Design:iphone:: #39 Modify wave shape

- 물결 정도 변경

JUNY0110 committed:
Merge branch 'design/39-progress_cell_layout' of https://github.com/Samsamhada/Giwazip-Client-App into design/39-progress_cell_layout

JUNY0110 commented on GiwazipClient.xcodeproj/project.pbxproj:
```diff
@@ -21,6 +21,8 @@
 		260A63E829710A7F004034AD /* SubCoordinator.swift in Sources */ = {isa = PBXBuildFile; fileRef = 260A63E729710A7F004034AD /* SubCoordinator.swift */; };
 		260A63EA29711307004034AD /* BaseCoordinator.swift in Sources */ = {isa = PBXBuildFile; fileRef = 260A63E929711307004034AD /* BaseCoordinator.swift */; };
 		260A63F22975E01F004034AD /* SplashViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 260A63F12975E01F004034AD /* SplashViewController.swift */; };
+		2614F24F2997DA1C0051EE51 /* ProgressCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2614F24E2997DA1C0051EE51 /* ProgressCell.swift */; };
+		2614F251299802BE0051EE51 /* WaveView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2614F250299802BE0051EE51 /* WaveView.swift */; };
```
p3.
파일 위치를 ASCell 위로 조정하는 게 어떨까요? 
뷰의 상단부터 하단순으로 Cell 위치를 조정하면 파일을 찾기 더 수월할 것 같습니다.

ikanic replied:
> p3. 파일 위치를 ASCell 위로 조정하는 게 어떨까요? 뷰의 상단부터 하단순으로 Cell 위치를 조정하면 파일을 찾기 더 수월할 것 같습니다.

8432abb 로 반영했습니다. 다만, 이렇게 변경했을 때 오히려 눈에 잘 안들어오는데, abc 순으로 변경하는 것은 어떨까요?

JUNY0110 replied:
> [8432abb](https://github.com/Samsamhada/Giwazip-Client-App/commit/8432abb3206fcfa9263f57492c104303af2dcea0) 로 반영했습니다. 다만, 이렇게 변경했을 때 오히려 눈에 잘 안들어오는데, abc 순으로 변경하는 것은 어떨까요?

화면이 다양하지 않아 폴더링으로 구분하는건 애매하니, 알파벳순으로라도 하면 규칙성이 생겨 보기에 낫겠네요.
현재는 화면단위로 파일을 모아두기만 했을뿐, 별다른 규칙성이 없어서 화면단위로 나눈 것처럼 보이지도 않기도 하구요.

ikanic replied:
> > [8432abb](https://github.com/Samsamhada/Giwazip-Client-App/commit/8432abb3206fcfa9263f57492c104303af2dcea0) 로 반영했습니다. 다만, 이렇게 변경했을 때 오히려 눈에 잘 안들어오는데, abc 순으로 변경하는 것은 어떨까요?
> 
> 화면이 다양하지 않아 폴더링으로 구분하는건 애매하니, 알파벳순으로라도 하면 규칙성이 생겨 보기에 낫겠네요. 현재는 화면단위로 파일을 모아두기만 했을뿐, 별다른 규칙성이 없어서 화면단위로 나눈 것처럼 보이지도 않기도 하구요.

fa0c3f0 에서 우선 셀 부분만 abc 순으로 정렬 변경했습니다. 삭제 예정인 빈 cell은 최상단에 우선 그냥 두었습니다. 감사합니다.

JUNY0110 replied:
> [fa0c3f0](https://github.com/Samsamhada/Giwazip-Client-App/commit/fa0c3f02c6dc52bbfcba94769f2742275cd5d4de) 에서 우선 셀 부분만 abc 순으로 정렬 변경했습니다. 삭제 예정인 빈 cell은 최상단에 우선 그냥 두었습니다. 감사합니다.

확인했습니다.

JUNY0110 commented on GiwazipClient/Cells/ProgressCell.swift:
```diff
@@ -0,0 +1,91 @@
+    // MARK: - View
+
+    private let view: UIView = {
+        $0.layer.borderColor = UIColor.black.cgColor
```
p2.
border는 default Color가 black이라 없어도 삭제해도 무방할 것 같습니다.

ikanic replied:
> p2. border는 default Color가 black이라 없어도 삭제해도 무방할 것 같습니다.

d5327f5 로 반영했습니다. 감사합니다.

JUNY0110 commented on GiwazipClient/Cells/ProgressCell.swift:
```diff
@@ -0,0 +1,91 @@
+
+    // MARK: - View
+
+    private let view: UIView = {
```
p3.
이름이 view라고 하니, 뷰의 역할을 알기 어렵네요.
progressBaseView, baseView, borderView처럼 어떤 역할을 갖는 뷰인지 명시해주면 좋을 것 같습니다.

ikanic replied:
> p3. 이름이 view라고 하니, 뷰의 역할을 알기 어렵네요. progressBaseView, baseView, borderView처럼 어떤 역할을 갖는 뷰인지 명시해주면 좋을 것 같습니다.

해당 부분은 baseView, progressView 같은 느낌보다는 Frame 자체를 잡기위해 존재하는 뷰이기 때문에 view로 잡았습니다.

JUNY0110 replied:
> 해당 부분은 baseView, progressView 같은 느낌보다는 Frame 자체를 잡기위해 존재하는 뷰이기 때문에 view로 잡았습니다.

view는 VC에서 흔히 사용하는 이름이라, 혼용되기 쉬울 것 같습니다. 최소 frameView정도의 이름을 부여하는 것은 어떨까요?

ikanic replied:
> > 해당 부분은 baseView, progressView 같은 느낌보다는 Frame 자체를 잡기위해 존재하는 뷰이기 때문에 view로 잡았습니다.
> 
> view는 VC에서 흔히 사용하는 이름이라, 혼용되기 쉬울 것 같습니다. 최소 frameView정도의 이름을 부여하는 것은 어떨까요?

264cd61 에서 view는 categoryFrame으로 silhouette은 unselectOpacity로 변경되었습니다. 감사합니다.

JUNY0110 replied:
> [264cd61](https://github.com/Samsamhada/Giwazip-Client-App/commit/264cd619f64c4ac25b1f7054e7b35acfe9952bc5) 에서 view는 categoryFrame으로 silhouette은 unselectOpacity로 변경되었습니다. 감사합니다.

확인했습니다.

JUNY0110 commented on GiwazipClient/Cells/ProgressCell.swift:
```diff
@@ -0,0 +1,91 @@
+
+    // MARK: - Life Cycle
+
+    override private init(frame: CGRect) {
```
p2.
private은 빠지는 것으로 회의되었던 것 같습니다~

ikanic replied:
> p2. private은 빠지는 것으로 회의되었던 것 같습니다~

직접 넣어준 부분은 아니고, 자동 완성이 저렇게 넣어줘서 미쳐 확인하지 못했네요. fa57a2d 로 반영했습니다.

JUNY0110 commented on GiwazipClient/Cells/WaveView.swift:
```diff
@@ -0,0 +1,33 @@
+    override func draw(_ rect: CGRect) {
+        let shape = UIBezierPath()
+
+        shape.move(to: CGPoint(x: 0, y: bounds.maxY / 100 * (100 - progress)))
+        if 1..<100 ~= progress {
+            shape.addQuadCurve(to: CGPoint(x: bounds.maxX / 2, y: bounds.maxY / 100 * (100 - progress)), controlPoint: CGPoint(x: bounds.midX - bounds.midX / 2, y: bounds.maxY / 100 * (100 - progress) - 3))
+            shape.addQuadCurve(to: CGPoint(x: bounds.maxX, y: bounds.maxY / 100 * (100 - progress)), controlPoint: CGPoint(x: bounds.midX + bounds.midX / 2, y: bounds.maxY / 100 * (100 - progress) + 3))
+        } else {
+            shape.addLine(to: CGPoint(x: bounds.maxX, y: bounds.maxY / 100 * (100 - progress)))
+        }
+        shape.addLine(to: CGPoint(x: bounds.maxX, y: bounds.maxY))
+        shape.addLine(to: CGPoint(x: 0, y: bounds.maxY))
+        shape.addLine(to: CGPoint(x: 0, y: bounds.maxY / 100 * (100 - progress)))
```
p2.
읽기가.. 힘드네요.. 줄바꿈이 필요할 것 같습니다.
가능한 100자 이내에서는 줄바꿈하는 컨벤션이 있죠!

ikanic replied:
> p2. 읽기가.. 힘드네요.. 줄바꿈이 필요할 것 같습니다. 가능한 100자 이내에서는 줄바꿈하는 컨벤션이 있죠!

코드 자체가 좀 복잡하긴 합니다. 해당 부분 e962eb7 로 반영했습니다.

JUNY0110 commented on GiwazipClient/Views/HistoryViewController.swift:
```diff
@@ -66,7 +71,12 @@ extension HistoryViewController: UICollectionViewDelegate, UICollectionViewDataS
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         // TODO: - 5에 게시물 데이터 갯수 반영
-        return section == 0 ? 1 : 2
+
+        if section == 0 {
+            return isWorkingView ? 12 : 1
+        }
+
```
굿. 간단하게 분기하는 역할로 좋네요

의견)
isWorkingView으로 정한 이유가 있으신가요? isConverted 같은 이름은 어떤가요?

ikanic replied:
> 굿. 간단하게 분기하는 역할로 좋네요
> 
> 의견) isWorkingView으로 정한 이유가 있으신가요? isConverted 같은 이름은 어떤가요?

`isConverted`는 무엇을 convert 한다는 것인지 명확해 보이지 않는 것 같습니다. 문의내역인지 시공내역인지 판단한다는 느낌에서 `isWorkingView`가 조금 더 명확한 것 같네요.

JUNY0110 replied:
> `isConverted`는 무엇을 convert 한다는 것인지 명확해 보이지 않는 것 같습니다. 문의내역인지 시공내역인지 판단한다는 느낌에서 `isWorkingView`가 조금 더 명확한 것 같네요.

뷰를 작동하고 있는지? 정도로 보였는데, workingView인지? 라는 의미였군요.
View이름이 Bool과 같이 사용하니 모호한 감이 있긴하네요. 확인했습니다

ikanic replied:
> > `isConverted`는 무엇을 convert 한다는 것인지 명확해 보이지 않는 것 같습니다. 문의내역인지 시공내역인지 판단한다는 느낌에서 `isWorkingView`가 조금 더 명확한 것 같네요.
> 
> 뷰를 작동하고 있는지? 정도로 보였는데, workingView인지? 라는 의미였군요. View이름이 Bool과 같이 사용하니 모호한 감이 있긴하네요. 확인했습니다

621dc8e f26c3cb 에서 workingView가 workView로 바뀜에 따라 isWorkView로 변수 이름이 변경되었습니다.

JUNY0110 replied:
> 
> [621dc8e](https://github.com/Samsamhada/Giwazip-Client-App/commit/621dc8e5314eaefa362e5c9f24dd17b37d620a89) [f26c3cb](https://github.com/Samsamhada/Giwazip-Client-App/commit/f26c3cbde823ab58475063341da4495c508c8b04) 에서 workingView가 workView로 바뀜에 따라 isWorkView로 변수 이름이 변경되었습니다.

확인했습니다.

JUNY0110 commented on GiwazipClient/Views/HistoryViewController.swift:
```diff
@@ -92,28 +102,44 @@ extension HistoryViewController: UICollectionViewDelegate, UICollectionViewDataS
+        } else if isWorkingView {
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ProgressCell.identifier, for: indexPath) as! ProgressCell
+            // TODO: - 서버에서 가져온 값으로 카테고리 이름 및 진행률 변경
+            cell.progress = Double((indexPath.item % 10 + 1) * 10)
```
Q.
서버에서 Float로 받는 것으로 알고 있는데, Double로 작성한 이유가 있을까요?
어차피 서버에서 변환해준다면, 처음부터 Float로 하는 건 어떨까요?

ikanic replied:
> Q. 서버에서 Float로 받는 것으로 알고 있는데, Double로 작성한 이유가 있을까요? 어차피 서버에서 변환해준다면, 처음부터 Float로 하는 건 어떨까요?

해당 부분은 현재 정수이기 때문에 변환시켜주기 위해 필요했던 부분으로, 서버가 바뀐 현재는 Double로 형변환이 필요하지 않을 것 같습니다. 또한 해당 부분은 서버에서 progress를 가져올 시 변경될 부분인데, 현재 퍼센트가 올라간다를 눈으로 볼 수 있게 남겨둔 부분입니다. 해당 부분은 후에 서버 연동에서 변경하겠습니다.

그 외에 Cell의 Progress를 Double로 설정한 이유는 WaveView를 보시면 아시겠지만, 최종적으로 필요한 것은 CGFloat인데, Double -> CGFloat 형 변환은 알아서 잘 되는 반면에, Float -> CGFloat은 별도의 처리과정이 필요하기 때문에 Double로 설정했습니다.

ikanic replied:
> Q. 서버에서 Float로 받는 것으로 알고 있는데, Double로 작성한 이유가 있을까요? 어차피 서버에서 변환해준다면, 처음부터 Float로 하는 건 어떨까요?

최종적으로 필요한 타입이 CGFloat이기 때문에 535f9e8 으로 해당 부분 Double에서 CGFloat으로 변경 반영하였습니다.

JUNY0110 replied:
> > Q. 서버에서 Float로 받는 것으로 알고 있는데, Double로 작성한 이유가 있을까요? 어차피 서버에서 변환해준다면, 처음부터 Float로 하는 건 어떨까요?
> 
> 최종적으로 필요한 타입이 CGFloat이기 때문에 [535f9e8](https://github.com/Samsamhada/Giwazip-Client-App/commit/535f9e857f66d7212eca0dca6c31c487354ec180) 으로 해당 부분 Double에서 CGFloat으로 변경 반영하였습니다.

확인했습니다.

JUNY0110 commented on GiwazipClient/Views/HistoryViewController.swift:
```diff
@@ -92,28 +102,44 @@ extension HistoryViewController: UICollectionViewDelegate, UICollectionViewDataS
+            cell.progress = Double((indexPath.item % 10 + 1) * 10)
+            cell.categoryName.text = "안방"
+
+            if indexPath.item == 0 {
+                cell.isSelected = true
+                collectionView.selectItem(at: indexPath, animated: true, scrollPosition: .init())
+            }
```
p3.
뷰 동작과 관련된 코드라고 생각되어지는데, didSelectItemAt에서 사용하는게 자연스러울 것 같습니다.

Q.
animated: false
scrollPosition: ___
해당 내용이 다른 값으로 변경되어도 동작이 동일한데, 이유를 아실까요??

ikanic replied:
> p3. 뷰 동작과 관련된 코드라고 생각되어지는데, didSelectItemAt에서 사용하는게 자연스러울 것 같습니다.
> 
> Q. animated: false scrollPosition: ___ 해당 내용이 다른 값으로 변경되어도 동작이 동일한데, 이유를 아실까요??

해당 부분은 초기값을 지정하는 부분으로 터치 동작과 직접 연관이 있지 않고, didSelectItemAt으로 뺄 수 있는 부분도 아니기에 현재와 같이 넣었습니다.

추가로 `isSelected`의 동작이 iOS 15부터 `selectItem`으로 대체된 상황이라 중복된 기능으로 보여 73409c4 에서 `isSelected` 삭제했습니다.

selectItem 메서드는 맨 처음 선택된 아이템을 지정해주는 코드입니다.
현재는 0번째 아이템이 선택되는 형태인데요. 저희 앱은 카테고리 12개가 한 화면에 나오는 형태라서 차이가 없습니다.
이 부분이 달라지는 경우는, 만약 컨텐츠 부분에서 특정 게시물이 처음부터 선택되었다고 생각했을 때, 맨 처음에 해당 게시물로 포커스가 이동되어 보여지도록 할 때는 차이를 볼 수 있겠네요.
다만 해당 부분은 어디까지나 예측이고, 실제 어떻게 적용되는지는 해봐야 알 것 같습니다. iOS 버전 이슈도 있어서 확실하게 동작을 확인해보기가 쉽지 않네요.

JUNY0110 reviewed (changes requested):
고생하셨습니다. 오랜만에 클라작업이셨네요! 베지에를 사용하는 작업이라 까다로웠을텐데 잘 해주셔서 감사합니다.

taek0622 committed:
Chore:wrench:: #39 Delete init bordercolor

- 기본 색상과 같은 borderColor 지정 구문 삭제

taek0622 committed:
Chore:wrench:: Delete private init

- init에서 접근지정자 삭제

taek0622 committed:
Merge branch 'design/39-progress_cell_layout' of https://github.com/Samsamhada/Giwazip-Client-App into design/39-progress_cell_layout

taek0622 committed:
Chore:wrench:: Change file sequence

- 파일 순서 변경

taek0622 committed:
Style:art:: Modify Line space

- 100자 이내로 줄 바꿈

taek0622 committed:
Chore:wrench:: Delete isSelected toggle

- isSelected를 직접 변경하는 코드 삭제
- iOS 15 이상에서는 isSelected 프로퍼티를 직접 변경하는 것을 권장하지 않음
- iOS 14 이하에서는 해당 부분 동작 안함

taek0622 committed:
Chore:wrench:: Modify workingview

- workingview에서 workview로 이름 변경

taek0622 committed:
Chore:wrench:: Modify cell sequence

- 셀 순서 변경

taek0622 committed:
Chore:wrench:: Modify isWorkingView

- isWorkingView에서 isWorkView로 변경된 파라미터 수정 빼먹은 부분 수정

taek0622 committed:
Chore:wrench:: Modify view name

- 일부 셀 구성 요소 이름 변경

taek0622 committed:
Chore:wrench:: Modify progress type

- progress 타입을 double에서 cgfloat으로 변경

JUNY0110 reviewed (approved):
빠른 반영 감사합니다^^
